File: solo-leveling-ui-service/src/main/java/com/sleepkqq/sololeveling/ui/exception/CustomErrorHandler.java
package com.sleepkqq.sololeveling.ui.exception;

import static com.sleepkqq.sololeveling.ui.localization.LocalizationMessage.ERROR_UNEXPECTED;

import com.sleepkqq.sololeveling.ui.localization.LocalizationException;
import com.sleepkqq.sololeveling.ui.model.UserData;
import com.sleepkqq.sololeveling.ui.service.TgAuthService;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.component.notification.Notification;
import com.vaadin.flow.server.ErrorEvent;
import com.vaadin.flow.server.ErrorHandler;
import java.util.Locale;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.context.MessageSource;

@RequiredArgsConstructor
public class CustomErrorHandler implements ErrorHandler {

  private final TgAuthService tgAuthService;
  private final MessageSource messageSource;

  @Override
  public void error(ErrorEvent errorEvent) {
    var locale = tgAuthService.findCurrentUser().map(UserData::getLocale).orElse(Locale.ENGLISH);
    Optional.ofNullable(UI.getCurrent())
        .ifPresent(c -> c.access(() -> Notification.show(messageSource.getMessage(
            errorEvent.getThrowable() instanceof LocalizationException l
                ? l.getLocalizationMessage().getPath()
                : ERROR_UNEXPECTED.getPath(),
            null,
            locale
        ))));
  }
}

File: solo-leveling-ui-service/src/main/kotlin/com.sleepkqq.sololeveling/ui/exception/CustomErrorHandler.kt
package com.sleepkqq.sololeveling.ui.exception

import com.sleepkqq.sololeveling.ui.localization.LocalizationException
import com.sleepkqq.sololeveling.ui.localization.LocalizationMessage
import com.sleepkqq.sololeveling.ui.service.TgAuthService
import com.vaadin.flow.component.UI
import com.vaadin.flow.component.notification.Notification
import com.vaadin.flow.server.ErrorEvent
import com.vaadin.flow.server.ErrorHandler
import org.slf4j.LoggerFactory
import org.springframework.context.MessageSource
import java.util.Locale

class CustomErrorHandler(
	private val tgAuthService: TgAuthService,
	private val messageSource: MessageSource
) : ErrorHandler {

	private val log = LoggerFactory.getLogger(CustomErrorHandler::class.java)

	override fun error(errorEvent: ErrorEvent) {
		log.error("Error occurred", errorEvent.throwable)

		val exception = errorEvent.throwable
		val locale = tgAuthService.findCurrentUser()
			?.locale
			?: Locale.ENGLISH

		UI.getCurrent()
			?.let { it.access { showExceptionNotification(exception, locale) } }
	}

	private fun showExceptionNotification(exception: Throwable, locale: Locale) {
		val path = if (exception is LocalizationException)
			exception.localizationMessage.path
		else
			LocalizationMessage.ERROR_UNEXPECTED.path

		Notification.show(messageSource.getMessage(path, null, locale))
	}
}
